compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -985,7 +985,180 @@ OMR::Power::TreeEvaluator::vstoreiEvaluator(TR::Node *node, TR::CodeGenerator *c
+
+   node->setRegister(resReg);
+
+   //Evaluate sum across vector operand
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::vspltisw, node, temp, 0);
+   generateTrg1Src2Instruction(cg, sumOp, node, tempRes, srcReg, temp); //Sum across each word element
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vsumsws, node, tempRes, tempRes, temp); //Sum word elements across entire vector operand
+
+   //Copy result into GPR
+   if (cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_PPC_P9))
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::mfvsrld, node, resReg, tempRes);
+   else
+   {
+      generateTrg1Src2ImmInstruction(cg, TR::InstOpCode::xxpermdi, node, tempRes, tempRes, temp, 3); //move sum to upper doubleword element of tempRes
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::mfvsrd, node, resReg, tempRes);
+   }
+
+   cg->stopUsingRegister(tempRes);
+   cg->stopUsingRegister(temp);
+   cg->decReferenceCount(firstChild);
+
+   return resReg;
+   }
+
+TR::Register *OMR::Power::TreeEvaluator::vreductionAddWordHelper(TR::Node *node, TR::CodeGenerator *cg, TR::DataType type)
+   {
+   TR::Node *firstChild = node->getFirstChild();
+
+   TR::Register *srcReg = cg->evaluate(firstChild);
+   TR::Register *resReg;
+
+   TR::Register *shiftReg = cg->allocateRegister(TR_VRF);
+   TR::Register *temp = cg->allocateRegister(TR_VRF);
+   TR::Register *tempGPR;
+   TR::Register *tempVSX;
+
+   //choose addOp and allocate registers based on data type
+   TR::InstOpCode::Mnemonic addOp;
+
+   if (type == TR::Int32)
+   {
+      addOp = TR::InstOpCode::vadduwm;
+      tempVSX = cg->allocateRegister(TR_VRF);
+      tempGPR = cg->allocateRegister();
+      resReg = tempGPR;
+   }
+   else if (type == TR::Float)
+   {
+      addOp = TR::InstOpCode::xvaddsp;
+      tempVSX = cg->allocateRegister(TR_FPR);
+      resReg = tempVSX;
+   }
+   else
+   {
+      TR_ASSERT(false, "cannot call vreductionAddWordHelper on vector type %s\n", type.toString());
+      return NULL;
+   }
+
+   node->setRegister(resReg);
+
+   //set shift amount to -32 bits. Since only the lowest 6 bits of shiftReg are used by vrld as an unsigned integer, this will result in a shift of 32 bits
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::vspltisw, node, shiftReg, -16);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vadduwm, node, shiftReg, shiftReg, shiftReg);
+
+   //Evaluate sum across vector operand by rotating and performing lanewise addition
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vrld, node, temp, srcReg, shiftReg);
+   generateTrg1Src2Instruction(cg, addOp, node, tempVSX, srcReg, temp);
+   generateTrg1Src2ImmInstruction(cg, TR::InstOpCode::xxpermdi, node, temp, temp, temp, 2);
+   generateTrg1Src2Instruction(cg, addOp, node, tempVSX, tempVSX, temp);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vrld, node, temp, temp, shiftReg);
+   generateTrg1Src2Instruction(cg, addOp, node, tempVSX, tempVSX, temp);
```

Comment:
I don't think this is right. It looks like it ends up like this: (In this example, AB means A+B not A*B)

```
srcReg  =    A    B    C    D
temp    =    B    C    D    A
tempVSX =  A+B  B+C  C+D  A+D
temp    =    D    A    B    C
tempVSX =  ABD  ABC  BCD  ACD
temp    =    A    B    C    D
tempVSX = AABD ABBC BCCD ACDD
```

Also, I think this can be changed to be done in 2 adds by permuting tempVSX instead.
```
srcReg  =    A    B    C    D
temp    =    B    C    D    A
tempVSX =  A+B  B+C  C+D  A+D
temp    =  C+D  A+D  A+B  B+C
tempVSX = ABCD ABCD ABCD ABCD
```